codemp/rd-rend2/glsl/ssao.glsl

```diff
@@ -8,95 +8,116 @@ void main()
-
-float random( const vec2 p )
-{
-  // We need irrationals for pseudo randomness.
-  // Most (all?) known transcendental numbers will (generally) work.
-  const vec2 r = vec2(
-    23.1406926327792690,  // e^pi (Gelfond's constant)
-     2.6651441426902251); // 2^sqrt(2) (Gelfond-Schneider constant)
-  //return fract( cos( mod( 123456789., 1e-7 + 256. * dot(p,r) ) ) );
-  return mod( 123456789., 1e-7 + 256. * dot(p,r) );  
-}
+//
+// AO Shader by Monsterovich :D
+//
 
-mat2 randomRotation( const vec2 p )
-{
-	float r = random(p);
-	float sinr = sin(r);
-	float cosr = cos(r);
-	return mat2(cosr, sinr, -sinr, cosr);
+float readDepth( in vec2 coord, in float znear, in float zfar ) {
+	return (2.0 * znear) / (zfar + znear - texture( u_ScreenDepthMap, coord ).x * (zfar - znear));	
 }
 
-float getLinearDepth(sampler2D depthMap, const vec2 tex, const float zFarDivZNear)
-{
-		float sampleZDivW = texture(depthMap, tex).r;
-		return 1.0 / mix(zFarDivZNear, 1.0, sampleZDivW);
+float compareDepths( in float depth1, in float depth2, in float znear, in float zfar  ) {
+	float diff = sqrt( clamp(1.0-(depth1-depth2) / ( u_SSAOSettings2.y /* aorange */ / (zfar - znear)),0.0,1.0) );
+	float ao = min(u_SSAOSettings.x /* aocap */,max(0.0,depth1-depth2-u_SSAOSettings2.z /* depthTolerance */) * u_SSAOSettings.z /* aoMultiplier */) * diff;
+	return ao;
 }
 
-float ambientOcclusion(sampler2D depthMap, const vec2 tex, const float zFarDivZNear, const float zFar)
+const float P1 = 0.70710678118; // sin,cos(pi/4)
+const float P2x = 0.92387953251; // cos(pi/8)
+const float P2y = 0.38268343236; // sin(pi/8)
+const float P3x = P2y; // cos(3*pi/8)
+const float P3y = P2x; // sin(3*pi/8)
+
+
+void main(void)
 {
-	float result = 0;
-
-	float sampleZ = zFar * getLinearDepth(depthMap, tex, zFarDivZNear);
-
-	vec2 expectedSlope = vec2(dFdx(sampleZ), dFdy(sampleZ)) / vec2(dFdx(tex.x), dFdy(tex.y));
-	
-	if (length(expectedSlope) > 5000.0)
-		return 1.0;
-	
-	vec2 offsetScale = vec2(3.0 / sampleZ);
-	
-	mat2 rmat = randomRotation(tex);
-		
+	float depth = readDepth( var_ScreenTex, u_ViewInfo.x, u_ViewInfo.y );
+
+	float d;
+ 
+	float pw = 1.0 / u_ScreenInfo.x;
+	float ph = 1.0 / u_ScreenInfo.y;
+ 
+	float ao = 0.0;
+	float aoScale = 1.0;
+
 	int i;
-	for (i = 0; i < 3; i++)
+	for (i = 0; i < 4; i++)
 	{
-		vec2 offset = rmat * poissonDisc[i] * offsetScale;
-		float sampleZ2 = zFar * getLinearDepth(depthMap, tex + offset, zFarDivZNear);
-
-		if (abs(sampleZ - sampleZ2) > 20.0)
-			result += 1.0;
-		else
-		{
-			float expectedZ = sampleZ + dot(expectedSlope, offset);
-			result += step(expectedZ - 1.0, sampleZ2);
-		}
+		// This creates a circle, using precalculated sin/cos for performance reasons
+		// pi / 8 (4 points)
+		d = readDepth( vec2(var_ScreenTex.x+pw*P2x,var_ScreenTex.y+ph*P2y), u_ViewInfo.x, u_ViewInfo.y);
+		ao += compareDepths(depth,d,u_ViewInfo.x,u_ViewInfo.y) / aoScale;
```

Comment:
Divisions are generally slower than multiplications.

This can be changed to..
```
ao += compareDepths(...) * aoScale;

...

aoScale *= 0.8;
```